### ISSUE TITLE
Type checking nonterminating function marked as total and used in type doesn't terminate

### ISSUE THREAD

**Issue by ionathanch** [idris-lang/Idris2#458]
Sorry if this is actually expected behaviour, or if this has already been reported!

# Steps to Reproduce

Type check this file (I used `idris2 bad.idr -c`):

```idris
total bad : Nat -> Nat
bad n = bad n

total bad_refl : (n: Nat) -> n = bad n
bad_refl n = Refl
```

# Expected Behavior

Type checker returns this error:

```
bad.idr:1:1--1:23:bad is not total:
	possibly not terminating due to recursive path Main.bad -> Main.bad at:
1	total bad : Nat -> Nat
	^^^^^^^^^^^^^^^^^^^^^^
```

# Observed Behavior

Type checker doesn't terminate. (Or at least appears not to. It sure uses up a lot of my memory.)

# More details

Curiously, this file *does* return the correct error:

```idris
total bad : Nat
bad = bad

total bad_refl : bad = bad
bad_refl = Refl
```

And so does this one:

```idris
total bad : Nat -> Nat
bad n = bad n

total bad_refl : (n: Nat) -> n = bad n
-- bad_refl n = Refl
```

**Comment by edwinb** [comment]
Indeed. This happens because we haven't yet told it not to reduce non-total things in types. Sorry about that...

I did want to think about it a little bit more, because there are situations where it's safe to reduce non total things that you want to reason about (say, things that can unfold once before diverging). But that's an exception, not the rule, so we shouldn't let that drive any other design decisions.
